### PR TITLE
Updating support of Megatron-LM

### PR DESCRIPTION
**Pull request by pengdurice**
# What does this PR do?

This PR is to make sure that accelerate continues support using Megatron-LM as backend for training large scale LLMs.
In details:
1. FP8 quantization of the model is by-passed if Megatron is used since the model will be re-initialized later on.
2. Added a set of useful parameters necessary  for enabling Megatron-LM: e.g. gradient checkpointing, cpu optimizer offloading.
3. Made changes on calling Megatron's functions since the func signatures have been changed.
4. Removed megatron_generate since it is no longer supported. 
5. Updated the user guide doc.
6. The change has been tested with fine-tuning GLM4.6 model using 64 H200 GPUs with 70K sequence length.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by pengdurice** [comment]
@SunMarc This is Peng. I am working on using accelerate and megatron-LM for fine tuning GLM4.6 models. This PR includes some updates of accelerate to better support new features as well as configurations from Megatron. Would you please help take a look? Thank you!

**Comment by SunMarc** [comment]
Of course, I will have a look this week !

**Comment by pengdurice** [comment]
> Of course, I will have a look this week !

Thank you! Take your time and have a nice holiday ahead if you are in the US!

File: src/accelerate/utils/dataclasses.py [hunk 1]
```diff
@@ -2568,9 +2625,25 @@ def __post_init__(self):
...
+            "recompute_granularity": self.recompute_granularity,
+            "recompute_method": self.recompute_method,
+            "recompute_num_layers": self.recompute_num_layers,
+            "attention_backend": self.attention_backend,
+            "expert_model_parallel_size": self.expert_model_parallel_size,
+            "context_parallel_size": self.context_parallel_size,
+            "attention_dropout": self.attention_dropout,
+            "hidden_dropout": self.hidden_dropout,
+            "attention_softmax_in_fp32": self.attention_softmax_in_fp32,
+            # "expert_tensor_parallel_size": self.expert_tensor_parallel_size,
```
Comment (SunMarc) [inline src/accelerate/utils/dataclasses.py]: don't need to pass it ?

**Review by SunMarc — APPROVED** [review]
Thanks a lot, we don't maintain that much megatron as it is quite complex for users but happy to have this PR !

File: src/accelerate/utils/dataclasses.py (hunk 1 above)
Comment (pengdurice) [reply src/accelerate/utils/dataclasses.py]: I don't quite find this config useful, so left it commented out. Let me uncomment it. thanks for the comment!

Comment (pengdurice) [reply src/accelerate/utils/dataclasses.py]: @SunMarc , fixed it with a new commit. Would you please take another look? Thanks!

**Comment by SunMarc** [comment]
@bot /style
